linalg/hypre.hpp

```diff
@@ -1560,10 +1560,10 @@ class HypreLOBPCG
    void Solve();
 
    /// Collect the converged eigenvalues
-   void GetEigenvalues(Array<double> & eigenvalues);
+   void GetEigenvalues(Array<double> & eigenvalues) const;
 
    /// Extract a single eigenvector
-   HypreParVector & GetEigenvector(unsigned int i);
+   const HypreParVector & GetEigenvector(unsigned int i) const;
```

Comment:
This is a "technically breaking" change (see the discussion [here](https://github.com/mfem/mfem/pull/2197#discussion_r624422647)). Perhaps we should keep the non-const versions with `MFEM_DEPRECATED`?

Reply:
I actually agree strongly with @YohannDudouit on this issue. In these classes the eigenvectors are computed and stored inside the HYPRE structs. Allowing users to modify these vectors feels like we're lending out our friend's car without their knowledge or permission. It just doesn't feel right. If I'm overruled we can keep the older versions but I think it's unwise.

Reply:
I don't think anyone will disagree that the right signature here is the new one.

The point about keeping the current non-const version as deprecated is being user-friendly. Instead of forcing users to change their code immediately upon upgrading to the new version, we just warn them, and let them do the update to the new const interface when they have the time (especially because it was _our_ "fault" we had the "wrong" signature to begin with).

Reply:
That's a good point, @v-dobrev. I'll add the deprecated member functions.

Reply:
It seems that keeping the non-const versions has an unintended consequence. We generally don't call `GetEigenvector` from a `const HypreLOBPCG` object so the compiler will generally choose the non-const version of these methods and display the deprecation wanring.  I have committed a work-around to avoid seeing the deprecation warning when we build our examples but users will see this warning even if they aren't modifying the eigenvectors (unless they use a similar work-around).

I guess we have two choices; 1) compel users to immediately stop modifying the internally stored eigenvectors or 2) issue a warning against something they may not even be doing and force them to implement a work-around or ignore the warning.

I'm fine either way. We can do whatever the reviewers think is best.

Reply:
Ah, yeah, that's unfortunate that the compiler will choose the non-const version if it is available.

The `const_cast` is pretty ugly just to avoid a false positive warning. My vote is to remove the non-const versions altogether unless there is a different better way that we haven't considered.

Reply:
One option is to use a different name for the const method like `ReadEigenvector` or `Eigenvector`. It is not optimal but it is not horrible either.

Reply:
I would be surprised if anyone is altering these vectors. I think it's worth the risk to remove the non-const versions and wait and see if anyone complains.

Reply:
I also think that if someone really wants to modify these vectors, he has to know what he is doing and do the `const_cast` on his side, that seems normal to me since these vectors are not meant to be altered.